qemu/tests/cpu_device_hotplug_maximum.py

```diff
@@ -0,0 +1,100 @@
+from virttest import utils_misc
+from virttest import utils_qemu
+from virttest.utils_version import VersionInterval
+
+from provider import cpu_utils
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Test hotplug maximum vCPU device.
+
+    1) Launch a guest without vCPU device.
+    2) Hotplug all vCPU devices and check successfully or not. (qemu side)
+    3) Check if the number of CPUs in guest changes accordingly. (guest side)
+    4) Reboot guest.
+    5) Hotunplug all vCPU devices and check successfully or not. (qemu side)
+
+    :param test:   QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env:    Dictionary with test environment.
+    """
+    os_type = params["os_type"]
+    machine_type = params["machine_type"]
+    mismatch_text = "Actual number of guest CPUs is not equal to the expected"
+    not_equal_text = "CPU quantity mismatched! Guest got %s but expected is %s"
+    # Many vCPUs will be plugged, it takes some time to bring them online.
+    verify_wait_timeout = params.get_numeric("verify_wait_timeout", 300)
+    qemu_binary = utils_misc.get_qemu_binary(params)
+    machine_info = utils_qemu.get_machines_info(qemu_binary)[machine_type]
+    machine_info = re.search(r'\(alias of (\S+)\)', machine_info)
+    current_machine = machine_info.group(1) if machine_info else machine_type
+    supported_maxcpus = (params.get_numeric("vcpu_maxcpus") or
+                         utils_qemu.get_maxcpus_hard_limit(qemu_binary,
+                                                           current_machine))
+    if not params.get_boolean("allow_pcpu_overcommit"):
+        supported_maxcpus = min(supported_maxcpus, cpu.online_cpus_count())
+
+    logging.info("Define the CPU topology of guest")
+    vcpu_devices = []
+    if (cpu.get_cpu_vendor_name() == "amd" and
+            params.get_numeric("vcpu_threads") != 1):
+        test.cancel("AMD cpu does not support multi threads")
+    elif machine_type.startswith("pseries"):
+        host_kernel_ver = uname()[2].split("-")[0]
+        if params.get_numeric("vcpu_threads") == 8:
+            supported_maxcpus -= divmod(supported_maxcpus, 8)[1]
+            vcpu_devices = ["vcpu%d" % c for c in
+                            range(1, supported_maxcpus // 8)]
+        # The maximum value of vcpu_id in 'linux-3.x' is 2048, so
+        # (vcpu_id * ms->smp.threads / spapr->vsmt) <= 256, need to adjust it
+        elif (supported_maxcpus > 256 and
+              host_kernel_ver not in VersionInterval("[4, )")):
+            supported_maxcpus = 256
+    vcpu_devices = vcpu_devices or ["vcpu%d" % vcpu for vcpu in
+                                    range(1, supported_maxcpus)]
+    params["vcpu_maxcpus"] = str(supported_maxcpus)
+    params["vcpu_devices"] = " ".join(vcpu_devices)
+    params["start_vm"] = "yes"
+
+    vm = env.get_vm(params["main_vm"])
+    vm.create(params=params)
+    session = vm.wait_for_login()
```

Comment:
if there many vcpus guest will spent lots of time to sync, it maybe take long time, so suggest to set timeout for wait_for_login function.

Reply:
Not needed here, guest only has 1 CPU before hotplug, but may take a long time during reboot phase if host's CPU is heavily overcommit. But it is not recommended for users, I would consider adding reboot timeout.